mmgen/models/architectures/dalle_mini/modules.py

```diff
@@ -0,0 +1,235 @@
+
+    def qkv(self, x):
+        """Calculate queries, keys and values for the embedding map.
+
+        Args:
+            x (torch.FloatTensor): Input feature map.
+
+        Returns:
+            q (torch.FloatTensor): Querie feature map.
+            k (torch.FloatTensor): Key feature map.
+            v (torch.FloatTensor): Value feature map.
+        """
+        q = self.querie(x)
+        k = self.key(x)
+        v = self.value(x)
+
+        return q, k, v
+
+    def forward(self, q, k, v, attention_mask):
+        """Forward function for attention.
+
+        Args:
+            q (torch.FloatTensor): Querie feature map.
+            k (torch.FloatTensor): Key feature map.
+            v (torch.FloatTensor): Value feature map.
+            attention_mask (torch.BoolTensor): whether to use
+                                                an attention mask.
+
+        Returns:
+            weights (torch.FloatTensor): Feature map after attention.
+        """
+        q = q.reshape(q.shape[:2] + (self.head_num, -1))
+        q /= q.shape[-1]**0.5
+        k = k.reshape(k.shape[:2] + (self.head_num, -1))
+        v = v.reshape(v.shape[:2] + (self.head_num, -1))
+
+        attention_bias = (1 - attention_mask.to(torch.float32)) * -1e12
+        weights = torch.einsum('bqhc,bkhc->bhqk', q, k)
+        weights += attention_bias
+        weights = torch.softmax(weights, -1)
+        weights = torch.einsum('bhqk,bkhc->bqhc', weights, v)
+        shape = weights.shape[:2] + (self.in_channels, )
+        weights = weights.reshape(shape)
+        weights = self.proj(weights)
+        return weights
+
+
+@MODULES.register_module()
+class EncoderLayer(nn.Module):
+    """EncoderLayer of the Bart model.
+
+    Ref:
+    https://github.com/kuprel/min-dalle/blob/main/min_dalle/models
+
+    Args:
+        in_channels (int): The channel number of the input feature map.
+        head_num (int): Number of heads in the attention.
+        out_channels (int): The channel number of the output feature map.
+    """
+
+    def __init__(self, in_channels, head_num, out_channels):
+        super().__init__()
+        self.selfAttention = AttentionBase(in_channels, head_num)
+        self.norm = build_norm_layer(dict(type='LN'), in_channels)[1]
+        self.glu = GLU(in_channels, out_channels)
+
+    def forward(self, x, attention_mask):
+        """Forward function for the encoder layer.
+
+        Args:
+            x (torch.FloatTensor): Input feature map.
+            attention_mask (torch.BoolTensor): whether to use
+                                            an attention mask.
+
+        Returns:
+            x (torch.FloatTensor): Output feature map.
+        """
+        residual = x.clone()
+        x = self.norm(x)
+        q, k, v = self.selfAttention.qkv(x)
+        x = self.selfAttention(q, k, v, attention_mask)
+        x = self.norm(x)
+        x = residual + x
+        residual = x.clone()
```

Comment:
In fact you can just write code in this way
```
        h = self.glu(x)
        x = h + x
```
instead of 
```
        residual = x.clone()
        x = self.glu(x)
        x = residual + x
```